parser-typechecker/src/Unison/Server/Endpoints/Definitions/Terms/TermSummary.hs

```diff
@@ -0,0 +1,140 @@
+instance ToCapture (Capture "fqn" Text) where
+  toCapture _ =
+    DocCapture
+      "fqn"
+      "The fully qualified name of a definition. The leading `.` is optional."
+
+instance ToSample TermSummary where
+  toSamples _ = noSamples
+
+data TermSummary = TermSummary
+  { fqn :: UnisonName,
+    hash :: UnisonHash,
+    summary :: DisplayObject SyntaxText SyntaxText,
+    tag :: TermTag
+  }
+  deriving (Generic, Show)
+
+instance ToJSON TermSummary where
+  toEncoding = genericToEncoding defaultOptions
+
+deriving instance ToSchema TermSummary
+
+-- TODO: Should this be using Backend.TermEntry as a way to get to TermSummary?
+serve ::
+  Var v =>
+  Handler () ->
+  Codebase IO v Ann ->
+  UnisonName ->
+  Maybe ShortBranchHash ->
+  Maybe NamespaceFQN ->
+  Maybe Width ->
+  Handler (APIHeaders TermSummary)
+serve tryAuth codebase rawTermName mayRoot namespaceName mayWidth =
+  let inBackend a = do
+        ea <- liftIO $ runExceptT a
+        errFromEither backendError ea
+
+      errFromEither f = either (throwError . f) pure
+
+      fqnToNamespacePath fqn = do
+        let fqnS = Text.unpack fqn
+        path' <- errFromEither (`Backend.BadNamespacePath` fqnS) $ parsePath' fqnS
+        pure (Path.fromPath' path')
+
+      width = mayDefaultWidth mayWidth
+
+      mkSummary reference termSig =
+        if Reference.isBuiltin reference
+          then BuiltinObject termSig
+          else UserObject termSig
+   in do
+        -- TODO: Should servant do this Name.unsafeFromString?
+        let termName = Name.unsafeFromText rawTermName
+
+        termSummary <- inBackend $ do
+          branch <- case namespaceName of
+            Nothing ->
+              Backend.resolveRootBranchHash mayRoot codebase
+            Just n -> do
+              namespacePath <- fqnToNamespacePath n
+              root <- Backend.resolveRootBranchHash mayRoot codebase
+              pure $ Branch.getAt' namespacePath root
+
+          let branch0 = Branch.head branch
+          let rel = Branch.deepTerms branch0
+          -- TODO: Map exception of missing element in set to 404 - noSuchDefinition
+          let termReferent = Set.elemAt 0 (Relation.lookupRan termName rel)
+          let termReference = Referent.toReference termReferent
+          let hash = Referent.toText termReferent
+          sig <- lift (Codebase.getTypeOfTerm codebase termReference)
+          case sig of
+            Nothing ->
+              throwError (Backend.MissingSignatureForTerm termReference)
+            Just typeSig -> do
+              hashLength <- liftIO $ Codebase.hashLength codebase
+              let ppe = Backend.basicSuffixifiedNames hashLength branch Path.empty
+              let formattedTermSig = Backend.formatSuffixedType (PrettyPrintEnvDecl ppe ppe) width typeSig
+              let summary = mkSummary termReference formattedTermSig
+              tag <- Backend.getTermTag branch0 termReferent sig
+
+              pure $ TermSummary rawTermName hash summary tag
```

Comment:
I feel like this is a lot of dense code to achieve this and could use some pointers to clean this up.